include/oneapi/dpl/pstl/hetero/dpcpp/execution_sycl_defs.h

```diff
@@ -42,21 +44,62 @@ struct DefaultKernelName;
 template <typename KernelName = DefaultKernelName>
 class device_policy
 {
+    // Needed for the copy constructor that rebinds the kernel name
+    template <typename>
+    friend class device_policy;
+
   public:
     using kernel_name = KernelName;
 
     device_policy() = default;
+
     template <typename OtherName>
-    device_policy(const device_policy<OtherName>& other) : q(other.queue())
+    device_policy(const device_policy<OtherName>& other)
+    {
+        ::std::scoped_lock lock{other.mtx};
+        q = other.q;
+    }
+
+    device_policy(const device_policy& other)
+    {
+        ::std::scoped_lock lock{other.mtx};
+        q = other.q;
+    }
+
+    device_policy(device_policy&& other)
     {
+        ::std::scoped_lock lock{other.mtx};
+        q.swap(other.q);
```

Comment:
Also header file for ```std::swap``` not included.

Reply:
I also don't understand why we do use `swap` member function for move and don't use it for copy. I feel like reusing copy/move-assignment from `optional` is the right way to go. But if you have arguments why you did `swap` please clarify.

My other question is can we make _member-init-list_ there instead of default constructing and then assigning? For example, a`static` function that that locks and returns optional with the correct value category?

> Also header file for std::swap not included.

That's something I don't understand at all. we don't use std::swap (at least in this line). For member function including `<optional>` is enough

Reply:
I mean if we take a loot into https://en.cppreference.com/w/cpp/algorithm/swap,
there are 3 headers in which ```std::swap``` declared:
```
Defined in header [<algorithm>](https://en.cppreference.com/w/cpp/header/algorithm) (until C++11)
Defined in header [<utility>](https://en.cppreference.com/w/cpp/header/utility) (since C++11)
Defined in header [<string_view>](https://en.cppreference.com/w/cpp/header/string_view) (since C++17)
```
And when I see introduced ```std::swap``` call I'm waiting for the inclusion of one from these headers.

Reply:
I switched the swap with the move constructor and rewrote the copy/move constructor as suggested by @rarutyun.